src/Microsoft.AspNetCore.WebHooks.Receivers/Filters/WebHookVerifyBodyTypeFilter.cs

```diff
@@ -97,118 +128,153 @@ public void OnResourceExecuting(ResourceExecutingContext context)
             }
 
+            var routeData = context.RouteData;
+            if (!routeData.TryGetWebHookReceiverName(out var receiverName))
+            {
+                return;
+            }
+
+            var bodyTypeMetadata = _bodyTypeMetadata;
+            if (_allBodyTypeMetadata != null)
+            {
+                var receiverBodyTypeMetadata = _allBodyTypeMetadata
+                    .FirstOrDefault(metadata => metadata.IsApplicable(receiverName));
+                if (receiverBodyTypeMetadata == null)
+                {
+                    // WebHookMetadataProvider requires an IWebHookBodyTypeMetadataService implementation for all
+                    // receivers. But, this case may occur if user has no receiver-specific actions i.e. uses only the
+                    // GeneralWebHookAttribute.
+                    _logger.LogCritical(
+                        0,
+                        "No '{MetadataType}' implementation found for the '{ReceiverName}' WebHook receiver. Each " +
+                        "receiver must register a '{ServiceMetadataType}' service.",
+                        typeof(IWebHookBodyTypeMetadataService),
+                        receiverName,
+                        typeof(IWebHookBodyTypeMetadataService));
+
+                    // Reuse the message for the Exception the WebHookMetadataProvider should have thrown.
+                    var message = string.Format(
+                        CultureInfo.CurrentCulture,
+                        Resources.Shared_MissingMetadata,
+                        typeof(IWebHookBodyTypeMetadataService),
+                        receiverName);
+                    throw new InvalidOperationException(message);
+                }
+
+                if (bodyTypeMetadata.BodyType == WebHookBodyType.All)
+                {
+                    // Use receiver-specific requirement since the action is flexible.
+                    bodyTypeMetadata = receiverBodyTypeMetadata;
+                }
+                else if ((~receiverBodyTypeMetadata.BodyType & bodyTypeMetadata.BodyType) != 0)
+                {
+                    // Failed subset check that WebHookMetadataProvider could not perform: Attribute must require the
+                    // same body type as receiver's metadata service or a subset. That is, `bodyTypeMetadata.BodyType`
+                    // flags must not include any beyond those set in `receiverBodyTypeMetadata.BodyType`.
+                    _logger.LogCritical(
+                        2,
+                        "Invalid '{MetadataType}.{PropertyName}' value '{PropertyValue}' in {AttributeType}. " +
+                        "This value must be equal to or a subset of the " +
+                        "'{ServiceMetadataType}.{ServicePropertyName}' value '{ServicePropertyValue}' for the " +
+                        "'{ReceiverName}' WebHook receiver.",
+                        typeof(IWebHookBodyTypeMetadata),
```

Comment:
These things that are `consts` should not be parameters in the log message because they are not data. They should just be in the string itself.

Reply:
Please review other log messages in this change and update as needed.

Reply:
My intent was to avoid manual changes if the type name changes. Are you recommending I change e.g.
``` c#
"'{ServiceMetadataType}.{ServicePropertyName}' value '{ServicePropertyValue}' for the "
```
to
``` c#
$"'{typeof(IWebHookBodyTypeMetadataService)}.{nameof(IWebHookBodyTypeMetadataService.BodyType)}' value '{{ServicePropertyValue}}' for the "
```